### DISCUSSION TITLE
Not working when using data obtained via PyAudio Callback

### DISCUSSION THREAD

**Discussion by michahn01** [dscripka/openWakeWord#102] [Q&A] [answered]
A question for a problem similar to mine had been posted before ([here](https://github.com/dscripka/openWakeWord/discussions/23)), but I'm having difficulties even after taking the steps prescribed in that discussion's solution. 

Mainly, my problem is that if I use audio data obtained via a stream_callback function of a PyAudio stream, I get an error about the argument to the model's prediction function not being a tuple. 

Here is my code:

```
import openwakeword
from openwakeword.model import Model
import pyaudio
import numpy as np

CHANNELS=1
CHUNK=256

model = Model(
    wakeword_models=["./hey_jarvis_v0.1.tflite"]
)

def callback(in_data, a=0, b=0, c=0):
    in_data = np.frombuffer(in_data, dtype=np.int16)
    frame = in_data.astype(np.int16)
    prediction = model.predict(frame)

    if prediction["hey_jarvis_v0.1"] > 0.7:
        print("Wakeword detected")

pa = pyaudio.PyAudio()
stream = pa.open(format=pyaudio.paInt16,
                 channels=CHANNELS,
                 rate=16000,
                 input=True,
                 frames_per_buffer=CHUNK,
                 stream_callback=callback)

try:
    while True:
        continue

except KeyboardInterrupt:
    stream.stop_stream()
    stream.close()
    pa.terminate()
    print("Stopped")
```

Here, I am doing `in_data = np.frombuffer(in_data, dtype=np.int16)`to convert the in_data buffer object, and I'm doing `frame = in_data.astype(np.int16)` to ensure the correct data type is being used. 

However, when I try running the code, I get the following error:

```
SystemError: new style getargs format but argument is not a tuple
Traceback (most recent call last):
  File "{file_path}/oww_test.py", line 31, in <module>
    continue
SystemError
```

Any help would be greatly appreciated. Thank you!

**Comment by dscripka** [comment, marked as answer]
It looks like there were just a few issues with the code, but after making some changes it seems to work correctly now:

```python
import openwakeword
from openwakeword.model import Model
import pyaudio
import numpy as np

CHANNELS=1
CHUNK=1280

model = Model(
    wakeword_models=["./hey_jarvis_v0.1.tflite"]
)

def callback(in_data, frame_count, time_info, status):
    frame = np.frombuffer(in_data, dtype=np.int16)
    prediction = model.predict(frame)

    if prediction["hey_jarvis_v0.1"] > 0.7:
        print("Wakeword detected")
    return (None, pyaudio.paContinue)

pa = pyaudio.PyAudio()
stream = pa.open(format=pyaudio.paInt16,
                 channels=CHANNELS,
                 rate=16000,
                 input=True,
                 frames_per_buffer=CHUNK,
                 stream_callback=callback)

try:
    while True:
        continue

except KeyboardInterrupt:
    stream.stop_stream()
    stream.close()
    pa.terminate()
    print("Stopped")

```

One issue was the callback function needs a return value (see the example [here](https://people.csail.mit.edu/hubert/pyaudio/docs/#example-callback-mode-audio-i-o)). And then the other problem was the chunksize; openWakeWord needs at minimum 1280 samples at 16 khz (80 ms) as input, but the `CHUNK` value was at 256.

Hopefully this works on your system now.

**Reply by michahn01** [reply]
It works perfectly, thank you!
